services/base.spec.js

```diff
@@ -414,30 +415,77 @@ describe('BaseService', function() {
       })
 
       it('overrides the logo', function() {
-        const expLogo =
-          'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI0MCIgaGVpZ2h0PSI0MCIgdmlld0JveD0iMTIgMTIgNDAgNDAiPgo8cGF0aCBmaWxsPSIjMzMzMzMzIiBkPSJNMzIsMTMuNGMtMTAuNSwwLTE5LDguNS0xOSwxOWMwLDguNCw1LjUsMTUuNSwxMywxOGMxLDAuMiwxLjMtMC40LDEuMy0wLjljMC0wLjUsMC0xLjcsMC0zLjIgYy01LjMsMS4xLTYuNC0yLjYtNi40LTIuNkMyMCw0MS42LDE4LjgsNDEsMTguOCw0MWMtMS43LTEuMiwwLjEtMS4xLDAuMS0xLjFjMS45LDAuMSwyLjksMiwyLjksMmMxLjcsMi45LDQuNSwyLjEsNS41LDEuNiBjMC4yLTEuMiwwLjctMi4xLDEuMi0yLjZjLTQuMi0wLjUtOC43LTIuMS04LjctOS40YzAtMi4xLDAuNy0zLjcsMi01LjFjLTAuMi0wLjUtMC44LTIuNCwwLjItNWMwLDAsMS42LTAuNSw1LjIsMiBjMS41LTAuNCwzLjEtMC43LDQuOC0wLjdjMS42LDAsMy4zLDAuMiw0LjcsMC43YzMuNi0yLjQsNS4yLTIsNS4yLTJjMSwyLjYsMC40LDQuNiwwLjIsNWMxLjIsMS4zLDIsMywyLDUuMWMwLDcuMy00LjUsOC45LTguNyw5LjQgYzAuNywwLjYsMS4zLDEuNywxLjMsMy41YzAsMi42LDAsNC42LDAsNS4yYzAsMC41LDAuNCwxLjEsMS4zLDAuOWM3LjUtMi42LDEzLTkuNywxMy0xOC4xQzUxLDIxLjksNDIuNSwxMy40LDMyLDEzLjR6Ii8+Cjwvc3ZnPgo='
         const badgeData = DummyService._makeBadgeData(
-          { logo: 'github', style: 'social' },
-          {}
+          { logo: 'github' },
+          { namedLogo: 'appveyor' }
```

Comment:
OK, so `logo` is more important than `namedLogo` here for...reasons.

I guess my feedback on this PR is broadly the same as the cache PR. Everything seems to work in staging, but we now have this non-trivial relationship between various ways we could inject a logo in different forms at different layers in the application. i.e: there is some relationship beween: `namedLogo`, `logo`, `defaultLogo`, `overrideLogo`, `logoSvg` etc and some of those override each other in various situations.

I haven't fully worked out what those relationships are from the code, but it should at least be explained somewhere so we can answer the question "if I pass `namedLogo`, `defaultLogo` and `logoSvg` all at the same time, which one is supposed to 'win'?".

Are you thinking of this as a permanent solution, or is it like the cache PR where as we refactor we can remove some of these layers?

Reply:
> I guess my feedback on this PR is broadly the same as the cache PR. Everything seems to work in staging, but we now have this non-trivial relationship between various ways we could inject a logo in different forms at different layers in the application. i.e: there is some relationship beween: `namedLogo`, `logo`, `defaultLogo`, `overrideLogo`, `logoSvg` etc and some of those override each other in various situations.
> 
> I haven't fully worked out what those relationships are from the code, but it should at least be explained somewhere so we can answer the question "if I pass `namedLogo`, `defaultLogo` and `logoSvg` all at the same time, which one is supposed to 'win'?".

Yea, that makes sense. I'll add a paragraph here.

I think a partial solution is to split out `_makeBadgeData` into its own file and to reorganize its tests so all the logo tests are together, all the label tests are together, all the color tests are together, etc., instead of now, where they're grouped by the kind of test they are. Between a descriptive paragraph, and reading a collection of tests all about logo, it should be clear to work out the desired behavior and reason about the correctness of the implementation.

> Are you thinking of this as a permanent solution, or is it like the cache PR where as we refactor we can remove some of these layers?

There are two things left that I think we can get rid of.

1. If we move named logo resolution into `makeBadge` then that part will be more contained. In particularly it'll be clear where the tests belong. The resolution and preparation logic will be tested around `makeBadge`, and only the overriding logic will need to be tested here. There won't be any less overriding logic, though it should be clearer whose job it is to deal with it and to ensure it's correct.
2. When the GitHub, Twitter, and Liberapay services are refactored, the legacy logo overriding pathway (`makeLogo`) will become an implementation detail of `makeBadgeData`. When the rest of the legacy services are gone, we'll be able to get rid of it entirely.

Reply:
I've added some docs about how the logo options are resolved.